src/encoding.rs

```diff
@@ -82,32 +82,38 @@ where
 ///
 /// Based loosely on [`ReadVarint64FromArray`][1].
 ///
+/// ## Safety
+///
+/// The caller must ensure that `bytes` is non-empty and either `bytes.len() >= 10` or the last
+/// element in bytes is < `0x80`.
+///
 /// [1]: https://github.com/google/protobuf/blob/3.3.x/src/google/protobuf/io/coded_stream.cc#L365-L406
 #[inline]
-fn decode_varint_slice(bytes: &[u8]) -> Result<(u64, usize), DecodeError> {
+#[cfg(not(target_pointer_width = "64"))]
+unsafe fn decode_varint_slice(bytes: &[u8]) -> Result<(u64, usize), DecodeError> {
     // Fully unrolled varint decoding loop. Splitting into 32-bit pieces gives better performance.
 
     let mut b: u8;
     let mut part0: u32;
-    b = bytes[0];
+    b = *bytes.get_unchecked(0);
```

Comment:
I'm fine with making these unchecked, but I do want to keep the bar high for introducing unsafety on the decoding side.  I'm pretty sure I benchmarked making these unchecked in the past and wasn't able to move the needle on the in-tree varint decoding benchmarks.  Did you measure improvements with this change in isolation?

Reply:
I did see a small but significant change, I'll try and get some good numbers here too.